C error with generics and functions

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V version**: V 0.2.2 c026d8b  (current git as of 2021-04-22)
**OS:** linux x86_64

**What did you do?**

This short program actually demonstrates 2 bugs, but I think both could be related 
to the type interference of function types.

```
fn neg(a int) int {
        return -a
}

fn indirect_call(func fn (int) int, a int) int {
        println(typeof(func).name)
        return func(a)
}

fn generic_call<T>(func T, a int) int {
        println(T.name)
        return func(a) // DOES NOT COMPILE: error: unknown function: func
}

fn generic_indirect_call<T>(func T, a int) int {
        println(T.name)
        return indirect_call(func, a)
}

fn main() {
        println(neg(1))
        println(indirect_call(neg, 2))
        // println(generic_call(neg, 3)) // DOES NOT COMPILE - see above
        println(generic_indirect_call(neg, 4)) // C error. This should never happen.
}
```


**What did you expect to see?**

Program compiles and runs.


**What did you see instead?**

```
==================
/tmp/v/bug.12754187955653504616.tmp.c:1547: error: ',' expected (got "func")
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

If you were not working with C interop, this is a compiler bug, please report the bug using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -2255,7 +2255,7 @@ pub fn (mut c Checker) fn_call(mut call_expr ast.CallExpr) ast.Type {
 	if !found {
 		if v := call_expr.scope.find_var(fn_name) {
 			if v.typ != 0 {
-				vts := c.table.get_type_symbol(v.typ)
+				vts := c.table.get_type_symbol(c.unwrap_generic(v.typ))
 				if vts.kind == .function {
 					info := vts.info as ast.FnType
 					func = info.func

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -371,18 +371,11 @@ fn (mut g Gen) fn_args(args []ast.Param, is_variadic bool) ([]string, []string) 
 		if arg_type_sym.kind == .function {
 			info := arg_type_sym.info as ast.FnType
 			func := info.func
-			if !info.is_anon {
-				g.write(arg_type_name + ' ' + caname)
-				g.definitions.write_string(arg_type_name + ' ' + caname)
-				fargs << caname
-				fargtypes << arg_type_name
-			} else {
-				g.write('${g.typ(func.return_type)} (*$caname)(')
-				g.definitions.write_string('${g.typ(func.return_type)} (*$caname)(')
-				g.fn_args(func.params, func.is_variadic)
-				g.write(')')
-				g.definitions.write_string(')')
-			}
+			g.write('${g.typ(func.return_type)} (*$caname)(')
+			g.definitions.write_string('${g.typ(func.return_type)} (*$caname)(')
+			g.fn_args(func.params, func.is_variadic)
+			g.write(')')
+			g.definitions.write_string(')')
 		} else {
 			s := '$arg_type_name $caname'
 			g.write(s)

--- a/vlib/v/tests/generics_with_generics_fn_type_parameter_test.v
+++ b/vlib/v/tests/generics_with_generics_fn_type_parameter_test.v
@@ -1,0 +1,34 @@
+fn neg(a int) int {
+	return -a
+}
+
+fn indirect_call(func fn (int) int, a int) int {
+	println(typeof(func).name)
+	return func(a)
+}
+
+fn generic_call<T>(func T, a int) int {
+	println(T.name)
+	return func(a)
+}
+
+fn generic_indirect_call<T>(func T, a int) int {
+	println(T.name)
+	return indirect_call(func, a)
+}
+
+fn test_generics_with_generics_fn_type_parameter() {
+	mut ret := 0
+
+	ret = indirect_call(neg, 2)
+	println(ret)
+	assert ret == -2
+
+	ret = generic_call(neg, 3)
+	println(ret)
+	assert ret == -3
+
+	ret = generic_indirect_call(neg, 4)
+	println(ret)
+	assert ret == -4
+}

